### PR TITLE
net: http_server: Fix capturing of request headers on concurrent HTTP2 streams

### LINKED ISSUE #82273: net: http_server: capturing request headers does not correctly account for concurrent HTTP2 streams
**Describe the bug**
The header capture feature of the HTTP server (`CONFIG_HTTP_SERVER_CAPTURE_HEADERS`) does not correctly handle HTTP2 concurrent streams. In the case of a dynamic POST request, headers associated with one request could sometimes be accessed by an application callback associated with a different request.

**To Reproduce**
I think it would actually be quite difficult to create this situation in "real" usage of the HTTP server, but the situation would be:
- Send two concurrent POST requests to the HTTP server, ideally to two separate paths (I can't think of a logical use-case for concurrent requests from the same client to the same path).
- The requests should be interleaved: HEADERS1, HEADERS2, DATA1, DATA2
- The dynamic resource callbacks for each request might be able to access headers associated with the other request. For example the callback sending DATA1 to the application happens after HEADERS2 has been parsed and stored in the client context.

Perhaps the easiest way to demonstrate this is the failing test case I've created in this branch: https://github.com/mrodgers-witekio/zephyr/tree/http2_request_headers_concurrent_streams

**Expected behavior**
Each request should call the dynamic response callback with the correct headers associated with that request.

**Impact**
Minimal, since concurrent HTTP2 POST requests with the header capture feature enabled and different headers in each request is unlikely to happen often. The issue shouldn't happen with concurrent GET requests, since the header frames of two different streams cannot be interleaved, only header frames with data frames.

**Additional context**
I didn't consider this situation when implementing the header capture feature, and only just noticed the issue now. I have a few ideas on how to fix, but it would be good to get some thoughts on the right way to go before I submit a PR.

Based on RFC9113 section 4.3 I *think* that headers from different streams cannot not be interleaved (ie. HEADERS_START_1, HEADERS_START2, CONTINUATION_1, ...). If my understanding is correct, it makes fixing the issue significantly simpler since we do not need to simultaneously store headers for multiple different streams.

So for HTTP/2 we could call the application callback as soon as the headers are parsed to pass the headers to the application, even if no data is sent yet. Then the headers can be cleared ready for a request on a different stream. Any subsequent callbacks providing data to the application would not contain the header data which was passed in the first callback.

Since the application callback currently just accesses the headers from the `client_ctx` struct this cannot currently work, so a different way of passing the headers to the application would be needed. Maybe an additional parameter to the callback for request headers, or combine the request headers and data into a `struct http_request_ctx` similar to the `struct htp_response_ctx` used to pass response data back in the other direction?

Both of these would require a change to the dynamic resource callback signature which would break existing applications. If we want to avoid this, then maybe adding a pointer to the headers in the `client_ctx` struct, which points to the existing `header_capture_ctx` if the headers are valid for a given callback, or is set to `NULL` if not.

### PR DESCRIPTION
**Pull request by mrodgers-witekio**
Fixes #82273.

Concurrent HTTP POST requests on different HTTP2 concurrent streams require that the client's header_capture_context is re-used to capture headers on a second stream before all of the body data has been received (and sent to the application) on the first stream.

As a result, any captured headers must be sent to the application callback before any headers can be received on a different stream. In practice this means that for HTTP2 the application callback is called for the first time on receiving a headers frame, before any data frames are received. All subsequent application callbacks will not include the request header data.

While this mechanism is not necessary for HTTP1, it is also updated to only send headers in the first application callback for consistency. However for HTTP/1 the first callback may also include body data in addition to the headers, since there isn't the same clear separation of headers and data into separate frames.

I've also made the `header_capture_ctx` part of the `http_client_ctx` regardless of whether `CONFIG_HTTP_SERVER_HEADER_CAPTURE` is defined, since otherwise it would have been difficult to manage all of the `#if defined(...)` needed, and replaced with `if(IS_ENABLED(...))` where needed (has the bonus that the included code is compile checked regardless of which Kconfig options are defined). The buffers however are defined to a size of zero if `CONFIG_HTTP_SERVER_HEADER_CAPTURE` is not enabled, so the impact on code/ram size is fairly minimal in case the option is disabled (<200 bytes each for the sample app on ARM Cortex M).

While working on this fix I also spotted a second issue: since the `current_detail` was stored at a client level rather than a http2 stream level, concurrent HTTP2 POST requests would overwrite the resource detail and lead to data being either associated with the wrong resource detail, or not associated with any resource detail at all (giving a 404 response). This is fixed in the first commit.

### CODE REVIEW

**Review by rlubos — COMMENTED** [review]
Looks good, thank you for the contribution. Note, that we now have another example utilizing HTTP server API (https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/prometheus), need to align it with the API changes too.

**Comment by mrodgers-witekio** [comment]
Rebased and updated prometheus sample

**Review by rlubos — DISMISSED** [review]
Thanks

**Comment by mrodgers-witekio** [comment]
Rebased again as I noticed some of the test changes were included in the wrong commit (so tests would have failed on first commit)

File: include/zephyr/net/http/server.h [hunk 1]
```diff
@@ -162,12 +170,28 @@ enum http_data_status {
...
 	const char *value; /**< Pointer to header value NULL-terminated string. */
 };
 
+/** @brief HTTP request context */
+struct http_request_ctx {
+	uint8_t *data;                          /** HTTP request data */
+	size_t data_len;                        /** Length of HTTP request data */
+	struct http_header *headers;            /** Array of HTTP request headers */
+	size_t header_count;                    /** Array length of HTTP request headers */
+	enum http_header_status headers_status; /** Status of HTTP request headers */
```
Comment (jukkar) [inline include/zephyr/net/http/server.h:192]: Not sure about doxygen comments, but do we need to set them `/**< ..... */` here (as they are in the same line as what is being documented)?

**Review by jukkar — DISMISSED** [review]

File: include/zephyr/net/http/server.h (hunk 1 above)
Comment (mrodgers-witekio) [reply include/zephyr/net/http/server.h:192]: Good point, I just copied the comment formatting from the `struct http_response_ctx` but it seems this is also wrong, and has comments associated with the wrong fields in the online docs.

Will update to fix both of these.

Comment (kartben) [reply include/zephyr/net/http/server.h:192]: it's correct this way :) it's the syntax for when the comment is put *after* the element it applies to, i.e. useful when one wants to make everything fit on a single line
https://www.doxygen.nl/manual/docblocks.html#memberdoc

Comment (kartben) [reply include/zephyr/net/http/server.h:192]: oops yeah sorry, looks like I'm blind today. @jukkar's suggestion IS the correct way :)

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**